imagetagger/imagetagger/images/models.py

```diff
@@ -68,10 +84,61 @@ class Meta:
         default=None
     )
     pinned_by = models.ManyToManyField(settings.AUTH_USER_MODEL, related_name='pinned_sets')
+    zip_finished = models.BooleanField(default=True)
 
     def root_path(self):
         return os.path.join(settings.IMAGE_PATH, self.path)
 
+    def zip_file_exists(self):
+        return os.path.isfile(self.zip_path())
+
+    def zip_path(self):
+        return os.path.join(settings.IMAGE_PATH, self.relative_zip_path())
+
+    def relative_zip_path(self):
+        return os.path.join(self.path, self.name) + '.zip'
+
+    def _update_zip_thread(self):
+        self.zip_finished = False
+        self.save()
```

Comment:
There is a update hole between checking if it is False and setting it False. Check it here in a transaction, or see if select for update would be ok.

Reply:
Definitely required to do something about atomicity here. As you are working on actual files, it might be feasible to use file-based locking instead of (or in addition to) database locking (e.g., using [fasteners](https://pypi.org/project/fasteners/)). However, be careful that the lock does not get inconsistent after a crash; it might prove useful to first do file-based locking, and after successfully acquiring that lock mark the zip as unfinished in the database.